core-bundle/src/Routing/AbstractPageRouteProvider.php

```diff
@@ -132,23 +133,14 @@ protected function compareRoutes(Route $a, Route $b, array $languages = null): i
         $langB = null;
 
         if (null !== $languages && $pageA->rootLanguage !== $pageB->rootLanguage) {
-            $langA = $languages[$pageA->rootLanguage] ?? null;
-            $langB = $languages[$pageB->rootLanguage] ?? null;
-
-            $primaryA = substr($pageA->rootLanguage, 0, 2);
-            $primaryB = substr($pageB->rootLanguage, 0, 2);
-
-            if ($primaryA !== $primaryB) {
-                // We must not compare by language (without region) if they are the same, otherwise
-                // two pages with same language but different regions might not be sorted by region priority.
-                if (null === $langA) {
-                    $langA = $languages[$primaryA] ?? null;
-                }
-
-                if (null === $langB) {
-                    $langB = $languages[$primaryB] ?? null;
-                }
-            } elseif (null === $langA && null === $langB) {
+            $fallbackA = LocaleUtil::getFallbacks($pageA->rootLanguage);
+            $fallbackB = LocaleUtil::getFallbacks($pageB->rootLanguage);
+            $langsA = array_intersect(array_diff($fallbackA, $fallbackB), array_keys($languages));
+            $langsB = array_intersect(array_diff($fallbackB, $fallbackA), array_keys($languages));
+            $langA = $languages[end($langsA)] ?? null;
+            $langB = $languages[end($langsB)] ?? null;
+
+            if ($langA === $langB) {
```

Comment:
@contao/developers Any idea how this could be done more efficiently?

Reply:
something like this?

```php
$langA = null;
$langB = null;
$fallbackA = LocaleUtil::getFallbacks($pageA->rootLanguage);
$fallbackB = LocaleUtil::getFallbacks($pageB->rootLanguage);
foreach (array_reverse($fallbackA) as $locale) {
	if (isset($languages[$locale]) !\in_array($locale, $fallbackB, true)) {
		$langA = $languages[$locale];
		break;
	}
}
foreach (array_reverse($fallbackB) as $locale) {
	if (isset($languages[$locale]) !\in_array($locale, $fallbackA, true)) {
		$langB = $languages[$locale];
		break;
	}
}
```

Reply:
not bad. What's the `array_reverse` for?

Reply:
because the later values in $fallbackA/B have precedence? At least I thought that reading your code...

Reply:
Maybe even self documenting with a function? (naming is just a guess + needs to be adjusted)
```php
$getUniqueLanguage = static function (array $locales, array $notIn) use ($languages): ?string {
    foreach (array_reverse($locales) as $locale) {
        if (null !== ($language = $languages[$locale] ?? null) && !\in_array($locale, $notIn, true)) {
            return $language;
        }
    }

    return null;
};

$fallbackA = LocaleUtil::getFallbacks($pageA->rootLanguage);
$fallbackB = LocaleUtil::getFallbacks($pageB->rootLanguage);

$langA = $getUniqueLanguage($fallbackA, $fallbackB);
$langB = $getUniqueLanguage($fallbackB, $fallbackA);
```

Reply:
like so?

```php
$compareLocales = static function (string $rootLocale, string $compareLocale) use ($languages): ?int {
	$compareFallbacks = LocaleUtil::getFallbacks($compareLocale);
	foreach (array_reverse(LocaleUtil::getFallbacks($rootLocale)) as $locale) {
		if (isset($languages[$locale]) && !\in_array($locale, $compareFallbacks, true)) {
			return $languages[$locale];
		}
	}
	return null;
};

$langA = $compareLocales($pageA->rootLanguage, $pageB->rootLanguage);
$langB = $compareLocales($pageB->rootLanguage, $pageA->rootLanguage);
```

Reply:
Updated in 8387c4bd40c97438c4819229ece221969a83c273